File: web/src/lib/bizdays.ts
/**
 * Business-day math over the vendored ANBIMA calendar.
 *
 * Holidays come exclusively from `shared/anbima-calendar.json` — a committed,
 * reviewed snapshot of the ANBIMA calendar (see plan 02-01). No algorithmic or
 * third-party holiday-computing package may ever be introduced here (C-03):
 * a system that silently answers from a different calendar than the one that
 * was reviewed and committed is worse than one that refuses to answer.
 *
 * This module implements the exact same reference algorithm as
 * `cli/apollo_cli/bizdays.py` — see plan 02-02's <interfaces> block for the
 * shared specification. The two implementations must never be allowed to
 * disagree silently; `shared/bizdays.testcases.json` is the single source of
 * test data proving they don't.
 */

import calendar from "../../../shared/anbima-calendar.json";

export const CALENDAR_START: string = calendar.start;
export const CALENDAR_END: string = calendar.end;

const HOLIDAYS: ReadonlySet<string> = new Set(calendar.holidays);

export class InvalidDateError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "InvalidDateError";
  }
}

export class CalendarRangeError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "CalendarRangeError";
  }
}

const ISO_DATE_RE = /^\d{4}-\d{2}-\d{2}$/;

/**
 * Parse and validate an ISO `YYYY-MM-DD` date string, rejecting malformed or
 * non-existent calendar dates. Does NOT check the vendored calendar range —
 * see `assertInRange`.
 */
function parseIsoDate(date: string): { iso: string; utcMillis: number } {
  if (typeof date !== "string" || !ISO_DATE_RE.test(date)) {
    throw new InvalidDateError(`Invalid date format: ${JSON.stringify(date)}`);
  }

  const [yearStr, monthStr, dayStr] = date.split("-");
  const year = Number(yearStr);
  const month = Number(monthStr);
  const day = Number(dayStr);

  const utcMillis = Date.UTC(year, month - 1, day);
  const rebuilt = new Date(utcMillis);

  if (
    rebuilt.getUTCFullYear() !== year ||
    rebuilt.getUTCMonth() !== month - 1 ||
    rebuilt.getUTCDate() !== day
  ) {
    throw new InvalidDateError(`Invalid calendar date: ${JSON.stringify(date)}`);
  }

  return { iso: date, utcMillis };
}

/**
 * Assert that an already-parsed ISO date string falls within
 * `[CALENDAR_START, CALENDAR_END]` (inclusive on both ends). Plain
 * lexicographic string comparison is correct here because all dates are
 * zero-padded ISO strings.
 */
function assertInRange(iso: string): void {
  if (iso < CALENDAR_START || iso > CALENDAR_END) {
    throw new CalendarRangeError(
      `Date ${iso} is outside the vendored calendar range [${CALENDAR_START}, ${CALENDAR_END}]`,
    );
  }
}

export function isBusinessDay(date: string): boolean {
  const { iso, utcMillis } = parseIsoDate(date);
  assertInRange(iso);

  const dayOfWeek = new Date(utcMillis).getUTCDay();
  if (dayOfWeek === 0 || dayOfWeek === 6) {
    return false;
  }
  return !HOLIDAYS.has(iso);
}

function pad(n: number, width: number): string {
  return String(n).padStart(width, "0");
}

function formatIso(year: number, month: number, day: number): string {
  return `${pad(year, 4)}-${pad(month, 2)}-${pad(day, 2)}`;
}

export function addBusinessDays(date: string, n: number): string {
  const { iso, utcMillis } = parseIsoDate(date);
  assertInRange(iso);

  if (n === 0) {
    return iso;
  }

  const step = n > 0 ? 1 : -1;
  let remaining = Math.abs(n);
  const cursor = new Date(utcMillis);

  while (remaining > 0) {
    cursor.setUTCDate(cursor.getUTCDate() + step);
    const cursorIso = formatIso(
      cursor.getUTCFullYear(),
      cursor.getUTCMonth() + 1,
      cursor.getUTCDate(),
    );
    assertInRange(cursorIso);

    if (isBusinessDay(cursorIso)) {
      remaining -= 1;
    }
  }

  return formatIso(cursor.getUTCFullYear(), cursor.getUTCMonth() + 1, cursor.getUTCDate());
}

/**
 * Strictly-after semantics: a date that is already a business day still
 * advances. Deliberately NOT `bizdays`' `Calendar.following()`
 * same-day-passthrough behavior — see plan 02-02's backstop truth.
 */
export function nextBusinessDay(date: string): string {
  return addBusinessDays(date, 1);
}

File: cli/apollo_cli/bizdays.py
"""Business-day math over the vendored ANBIMA calendar.

Holidays come exclusively from the vendored `apollo_cli/data/anbima-calendar.json`
copy — a committed, reviewed snapshot of the ANBIMA calendar (see plan 02-01),
read via `importlib.resources` so it works identically whether this package is
an editable dev install or a real installed wheel with no monorepo checkout
anywhere on disk (see plan 24-01). `cli/tests/test_calendar_vendored_parity.py`
proves this vendored copy stays byte-identical to `shared/anbima-calendar.json`,
the original the `web/` app also reads. No algorithmic or third-party
holiday-computing package may ever be introduced here (C-03): a system that
silently answers from a different calendar than the one that was reviewed and
committed is worse than one that refuses to answer.

FORBIDDEN/PROHIBITED: `Calendar.load("ANBIMA")` anywhere in this module (and in
any shipped code). It sources holidays from the `bizdays` library's own bundled
table instead of the vendored JSON, violating CAL-03 and guaranteeing silent
drift on any future `bizdays` upgrade. The only correct calendar construction
is `Calendar(holidays=<vendored holidays>, weekdays=["Saturday", "Sunday"])`.

This module implements the exact same reference algorithm as
`web/src/lib/bizdays.ts` — see plan 02-02's <interfaces> block for the shared
specification. The two implementations must never be allowed to disagree
silently; `shared/bizdays.testcases.json` is the single source of test data
proving they don't.
"""

from __future__ import annotations

import json
import re
from datetime import date, timedelta
from importlib import resources
from typing import Final

from bizdays import Calendar

_ISO_DATE_RE: Final[re.Pattern[str]] = re.compile(r"^\d{4}-\d{2}-\d{2}$")

_CALENDAR_RESOURCE: Final = resources.files("apollo_cli.data").joinpath("anbima-calendar.json")
# Let a missing/unparseable file raise loudly at import time — never fall
# back to a bundled or algorithmic calendar.
_PAYLOAD: Final[dict[str, object]] = json.loads(_CALENDAR_RESOURCE.read_text(encoding="utf-8"))

CALENDAR_START: Final[str] = str(_PAYLOAD["start"])
CALENDAR_END: Final[str] = str(_PAYLOAD["end"])

_holidays_raw: object = _PAYLOAD["holidays"]
if not isinstance(_holidays_raw, list):
    msg = f"{_CALENDAR_RESOURCE}: 'holidays' is not a list"
    raise TypeError(msg)
_HOLIDAYS: Final[list[str]] = _holidays_raw

_CALENDAR: Final[Calendar] = Calendar(
    holidays=_HOLIDAYS,
    weekdays=["Saturday", "Sunday"],
)


class InvalidDateError(ValueError):
    """Raised for malformed or non-existent `YYYY-MM-DD` date strings."""


class CalendarRangeError(ValueError):
    """Raised for a date outside `[CALENDAR_START, CALENDAR_END]`."""


def _parse_iso_date(value: str) -> date:
    """Parse and validate an ISO `YYYY-MM-DD` date string, rejecting malformed
    or non-existent calendar dates. Does NOT check the vendored calendar
    range — see `_assert_in_range`.
    """
    if not isinstance(value, str) or not _ISO_DATE_RE.match(value):
        msg = f"Invalid date format: {value!r}"
        raise InvalidDateError(msg)

    try:
        return date.fromisoformat(value)
    except ValueError as exc:
        msg = f"Invalid calendar date: {value!r}"
        raise InvalidDateError(msg) from exc


def _assert_in_range(iso: str) -> None:
    """Assert that an ISO date string falls within
    `[CALENDAR_START, CALENDAR_END]` (inclusive on both ends). Plain
    lexicographic string comparison is correct here because all dates are
    zero-padded ISO strings.
    """
    if iso < CALENDAR_START or iso > CALENDAR_END:
        msg = (
            f"Date {iso} is outside the vendored calendar range [{CALENDAR_START}, {CALENDAR_END}]"
        )
        raise CalendarRangeError(msg)


def is_business_day(value: str) -> bool:
    parsed = _parse_iso_date(value)
    iso = parsed.isoformat()
    _assert_in_range(iso)
    return bool(_CALENDAR.isbizday(iso))


def add_business_days(value: str, n: int) -> str:
    parsed = _parse_iso_date(value)
    iso = parsed.isoformat()
    _assert_in_range(iso)

    if n == 0:
        return iso

    step = 1 if n > 0 else -1
    remaining = abs(n)
    cursor = parsed

    while remaining > 0:
        cursor = cursor + timedelta(days=step)
        cursor_iso = cursor.isoformat()
        _assert_in_range(cursor_iso)

        if is_business_day(cursor_iso):
            remaining -= 1

    return cursor.isoformat()


def next_business_day(value: str) -> str:
    """Strictly-after semantics: a date that is already a business day still
    advances. Deliberately NOT `bizdays`' `Calendar.following()`
    same-day-passthrough behavior — see plan 02-02's backstop truth.
    """
    return add_business_days(value, 1)
